### PR TITLE
 Clock and stopper classes + tests for each one of them (not all tests covered)

### PR DESCRIPTION
**Pull request by Afek-Sakaju**

### CODE REVIEW

File: classes/clock/clock.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ class Clock extends Time {
         if (this.interval) return;
 
         this.interval = setInterval(() => {
-            if (this.totalSeconds === MAX_CLOCK_SECONDS) this.pause();
+            if (this.tSeconds === MAX_CLOCK_SECONDS) this.pause();
```
Comment (hdriel) [inline classes/clock/clock.js:29]: I prefer first to add seconds 
then check if this.tSeconds >= MAX_CLOCK_SECONDS if so, pause 

because the clock can be not valid after to added one seconds last

File: classes/clock/__tests__/clock.test.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            test('clock: 23:59:50 auto pauses after reaching "23:59:59"', () => {
+                const clock = new Clock({
+                    seconds: 50,
+                    minutes: 59,
+                    hours: 23,
+                });
+                const maxLimit = '23:59:59';
+
+                setTimeout(() => {
+                    expect(clock.toString()).toBe(maxLimit);
```
Comment (hdriel) [inline classes/clock/__tests__/clock.test.js:59]: this line can't be reached 
please try to use it in another way 
try to use done param from test like 

```
test('clock: 23:59:50 auto pauses after reaching "23:59:59"', (done) => {
setTimeout(() => {
       expect(clock.toString()).toBe(maxLimit);
       done()
 }, 30000);
}
```

File: classes/clock/__tests__/clock.test.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    describe('valid cases', () => {
+        describe('start & pause methods tests', () => {
+            /*beforeEach(()=>{
+                    jest.runAllTimers();
+
+            *** not working that way, only after each setTimeout manually
+            
+            })*/
+
+            jest.useFakeTimers();
```
Comment (hdriel) [inline classes/clock/__tests__/clock.test.js:13]: I dont know about jest.useFakeTimers and jest.runAllTimers
I would happy to learn about that from you 😊

File: classes/stopper/__tests__/stopper.test.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            jest.useFakeTimers();
+
+            test.each([
+                [{ seconds: 0, minutes: 0, hours: 2 }, '02:00:00'],
+                [{ seconds: 0, minutes: 30, hours: 0 }, '00:30:00'],
+                [{ seconds: 45, minutes: 0, hours: 0 }, '00:00:45'],
+                [{ seconds: 10, minutes: 30, hours: 1 }, '01:30:10'],
+            ])(
+                'stopper of %s units, autoStart set to false then after 15 seconds returns %s',
+                (params, result) => {
```
Comment (hdriel) [inline classes/stopper/__tests__/stopper.test.js:22]: I've not tried to use done param with test.each , 
you are welcome to explore that and update me :)

File: classes/stopper/stopper.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            if (this.tSeconds === MIN_STOPPER_SECONDS) this.pause();
+            else this.subSeconds(1);
+        }, 1000);
+    }
+
+    pause() {
+        clearInterval(this.interval);
+    }
+
+    say() {
```
Comment (hdriel) [inline classes/stopper/stopper.js]: what is for?

File: classes/stopper/stopper.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            this.tSeconds = MIN_STOPPER_SECONDS;
+        }
+    }
+
+    start() {
+        if (this.interval) return;
+
+        this.interval = setInterval(() => {
+            if (this.tSeconds === MIN_STOPPER_SECONDS) this.pause();
+            else this.subSeconds(1);
```
Comment (hdriel) [inline classes/stopper/stopper.js]: same comment as clock class as comment

File: classes/stopper/stopper.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    ) {
+        super({ seconds, minutes, hours });
+
+        this.interval = null;
+
+        if (autoStart) this.start();
+    }
+
+    validateLimiter() {
+        if (this.tSeconds > MAX_STOPPER_SECONDS) {
```
Comment (hdriel) [inline classes/stopper/stopper.js:21]: why did you change the `this.totalSeconds` to `this.tSeconds` ?

File: classes/stopper/stopper.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
+const Time = require('../time/time');
+const {
+    MAX_STOPPER_SECONDS,
+    MIN_STOPPER_SECONDS,
+} = require('../stopper/utils/consts');
+
+// check what's the time limits of stopper?
+
+class Stopper extends Time {
```
Comment (hdriel) [inline classes/stopper/stopper.js:7]: in the stopper class, should have also stop method , 
todo: implement stop method

File: classes/clock/clock.js [hunk 9]
```diff
@@ -27,7 +27,7 @@ class Clock extends Time {
...
             else this.addSeconds(1);
```
Comment (hdriel) [inline classes/clock/clock.js]: I recommend to use `super.addSeconds(1)` although the **this** will work as well 
I think it's nicer to see where this method came from the inheritance..

**Review by hdriel — DISMISSED** [review]

File: classes/clock/clock.js (hunk 1 above)
Comment (Afek-Sakaju) [reply classes/clock/clock.js:29]: statement changed to stop clock as it reaches limit.
adding 1 second happens before the statement so the limit, wont break

File: classes/clock/__tests__/clock.test.js (hunk 2 above)
Comment (Afek-Sakaju) [reply classes/clock/__tests__/clock.test.js:59]: hadriel after i checked again, this line reached in the code
because i added before all tests:
`jest.useFakeTimers();`

and in each test the code: 
`jest.runAllTimers();`

File: classes/clock/__tests__/clock.test.js (hunk 3 above)
Comment (Afek-Sakaju) [reply classes/clock/__tests__/clock.test.js:13]: **Ty boss, its a big respect for me to teach you**

File: classes/stopper/__tests__/stopper.test.js (hunk 4 above)
Comment (Afek-Sakaju) [reply classes/stopper/__tests__/stopper.test.js:22]: ok ill check that sir

File: classes/clock/clock.js (hunk 9 above)
Comment (afekTheMiniLearner) [reply classes/clock/clock.js:31]: so should i always use super in this cases?

File: classes/stopper/stopper.js (hunk 7 above)
Comment (Afek-Sakaju) [reply classes/stopper/stopper.js:21]: because there is a get method at the Time class called totalSeconds

File: classes/clock/clock.js (hunk 9 above)
Comment (Afek-Sakaju) [reply classes/clock/clock.js]: changed
